File: 2020-03/3-29/46全排列/46.py
#  -*- coding: utf-8字符串转换整数 -*-

#  Editor      : Pycharm
#  File        : 46
#  Author      : ZhenyuMa
#  Created     : 2020/3/30 12:00 上午
#  Description : 回溯


from typing import List


class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        self.res = []
        self.visited = [False] * len(nums)
        self.helper(nums, [])
        return self.res

    def helper(self, nums, cur_res):
        if len(cur_res) == len(nums):
            self.res.append(cur_res[:])
            return

        for i in range(len(nums)):
            if not self.visited[i]:
                cur_res.append(nums[i])
                self.visited[i] = True
                self.helper(nums, cur_res)
                cur_res.pop()
                self.visited[i] = False


if __name__ == "__main__":
    test = Solution()
    print(test.permute([1, 2, 3]))

File: 2020-05/5-01/46全排列/46.cpp


class Solution {
private:
    vector<vector<int>> res;
public:
    vector<vector<int>> permute(vector<int>& nums) {
        vector<bool> visited = vector<bool>(nums.size(), false);
        helper(nums, vector<int>(), visited);
        return res;
    }

    void helper(const vector<int>& nums, vector<int> cur_res, vector<bool> visited){
        if(cur_res.size() == nums.size()){
            res.push_back(cur_res);
            return;
        }
        for(int i = 0; i < nums.size(); i++){
            if(!visited[i]){
                visited[i] = true;
                cur_res.push_back(nums[i]);
                helper(nums, cur_res, visited);
                visited[i] = false;
                cur_res.pop_back();
            }
        }
        return;
    }
};